Add ZAP Developer Docs (#268)
zaproxy/zaproxy-website: site/content/docs/developer/quick-start-build.md

---
title: "A Quick Start Guide to Building ZAP"
description: "This guide will get you started with running ZAP and installing add-ons from source."
images:
- https://www.zaproxy.org/img/docs/developer/zap-dev-build.png
tags: 
- guide
- tutorial
type: page
date: "2020-12-04"
addSocialPreview: true
---

Have you ever wanted to play around with the ZAP codebase but felt it was too overwhelming? This guide will walk you through building ZAP from the command line regardless of the operating system and IDE you're using. Let’s jump right into it!

## Preparation
All you need is a working local installation of [git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git) and [JDK 8+](https://adoptopenjdk.net/).

## Clone the Repo Locally
Create a new directory on your system and run the following command in that directory to clone the core ZAP repository.
```bash

> git clone https://github.com/zaproxy/zaproxy.git

Reviewer: Shouldn't people be cloning their fork as the first step? (So that origin properly points to their fork, and they don't end-up with origin/upstream the same).

Reviewer 2: I wanted to to make it as easy as possible for people to get ZAP running from the source code.
Cloning repos complicates things and isnt necessary. But happy to have some text pointing out that if people do plan to contribute back then they should go to step X

Author: Well, to prevent that we have a section at the end where they point origin to their fork (using `git remote set-url origin https://github.com/USERNAME/zaproxy.git`) and upstream to the main repo. But if cloning would be easier / more clear I can update the docs. What do you suggest?

Reviewer: @psiinon if these are developer docs isn't the assumption that they're contributing?

Or should we put this under a heading other than "developer"? (`/docs/desktop/run-from-source/`?)

Reviewer 2: Tricky - I want us to have a new section for ZAP internals - this is just the start of that section. But you are right, this page isnt just for ZAP devs. So open to suggestions ...
EG would love to migrate a load of this content: https://github.com/zaproxy/zaproxy/wiki/InternalDetails

Reviewer: Well we all know I over think these things so maybe we leave it as-is since there is a chunk later in the doc that addresses origin/upstream.

If we end-up with lots of questions or devs with problems we can always revisit.

Reviewer 3: Being developer does not imply being a contributor, e.g. you might want to create a custom version of ZAP for yourself. (There are many forks of zaproxy with custom changes that are not contributed back, offline changes should be a higher number I guess.)

Reviewer: Good point. Okay let's leave it as-is for now.

```

## Running ZAP from Source

> Navigate to the cloned *zaproxy* folder and run the following command (if you are a Windows user, omit the `./`):

Reviewer: On windows wouldn't you have to use `gradlew.bat`?

Reviewer 2: I think it will work without, but dont have a windows box to hand to check right now ;)

```bash
./gradlew run
```
That’s it! You’re now running ZAP from source! You can confirm this if you see “Dev Build” in the loading window.

> ![ZAP Dev Build Loading Window](/img/docs/developer/zap-dev-build.png)

Reviewer: Might be better to use a green arrow or green box, the blue looks like ZAP blue and part of the UI.

It is possible that you might be greeted with the following prompt if you already have some other application that is using port 8080.

![ZAP Proxy Port in Use](/img/docs/developer/zap-proxy-port.png)

> In this case, simply specify a different port and ZAP should start listening on that port.

Reviewer: The one that is being suggested should be free IIRC (it loops 10 times trying to find a free one).

ZAP uses a [gradle wrapper](https://docs.gradle.org/current/userguide/gradle_wrapper.html) that downloads all the dependencies for the project. The gradle wrapper has been configured to provide other tasks for convenience. You can view these by running,

```bash
./gradlew tasks
```

Note that while you have successfully got ZAP running, it doesn’t come with many of the key features that are provided by ZAP add-ons.

> ## Installing ZAP add-ons

Reviewer: ```suggestion
## Installing ZAP Add-ons
```

> As the name suggests, ZAP add-ons are a way to enhance ZAP functionality. There are several add-ons that allow you to do many different things. The *zap-extensions* repository contains the source code for these add-ons. So, the first thing to do is clone *zap-extensions*:

Reviewer: ```suggestion
As the name suggests, ZAP add-ons are a way to enhance ZAP functionality. There are several add-ons that allow you to do many different things. The *zap-extensions* repository contains the source code for the add-ons maintained by the core team. So, the first thing to do is clone *zap-extensions*:
```

There are third party add-ons maintained elsewhere.

```bash

> git clone https://github.com/zaproxy/zap-extensions.git.

Reviewer: Same comment re fork/origin/upstream.

Reviewer 2: Ditto ;)

```
Then, navigate to the cloned folder and run
```bash
./gradlew tasks
```
to see a list of the available tasks. You can observe many of these in the output, but this is the one we're interested in:
```
installZapAddOn - Installs the add-on into ZAP, listening on 8080 by default.
```
```bash
./gradlew installZapAddOn 
```
For this command to work as expected, you need to have an instance of ZAP running and listening on port 8080. If ZAP is running on a different port for you, then you can modify the above command as,
```bash
./gradlew installZapAddOn --port=1234
```

Note that if you run this task, ALL the ZAP add-ons in *zap-extensions* will be installed. To install a specific add-on, run the following command
```bash
./gradlew addOns:IDENTIFIER:installZapAddOn
```
where `IDENTIFIER` is the name of the add-on you want to install. You can find the identifier by looking at the structure of the *zap-extensions* repository:
```
├── zap-extensions
	├── addOns 
	│   ├── accessControl 
	│   ├── alertFilters 
	│   ├── alertReport 
	│   ├── allinonenotes 
	│   ├── amf 
	│   ├── ascanrules
```

> So, for instance, if you want to install the active and passive scan rules, you would run:

Reviewer: ```suggestion
So, for instance, if you want to install the Release status Active and Passive scan rules, you would run:
```

```bash
./gradlew addOns:ascanrules:installZapAddOn
./gradlew addOns:pscanrules:installZapAddOn
```
or if you're in a hurry,
```bash
./gradlew aO:ascanrules:iZAO
./gradlew aO:pscanrules:iZAO
```

The process of installing the ZAP HUD and the help is identical:
1. Clone their repositories.
    ```bash
    git clone https://github.com/zaproxy/zap-hud.git

>     git clone https://github.com/zaproxy/zap-core-help.git

Reviewer: fork/origin/upstream

    ```
2. Navigate to the cloned folders and run
	```bash
    ./gradlew installZapAddOn
    ```
For a better understanding of Gradle tasks you can check out the [Gradle user manual](https://docs.gradle.org/current/userguide/userguide.html).

## For Prospective ZAP Contributors
If you feel that some part of ZAP can be improved, enhanced or updated in any way, go ahead and open a [pull request](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/about-pull-requests)! Here are the steps you need to follow before you can do that:
1. Clone the main repositories as illustrated above.
1. Create a [GitHub](https://github.com) account.
1. Fork the ZAP repositories.
1. Update the tracked (remote) repositories.

> 1. Make changes and submit a PR!

Reviewer: I'd suggest swapping or clarifying the fork and clone steps.

### Fork the Repos on GitHub
Go to the GitHub page of each repository you cloned and [fork it](https://guides.github.com/activities/forking/) to your account (the button is towards the top right of the screen).

> ![The fork button on GitHub](/img/docs/developer/github-fork.png)

Reviewer: Same arrow comment

- https://github.com/zaproxy/zaproxy
- https://github.com/zaproxy/zap-extensions
- https://github.com/zaproxy/zap-hud
- https://github.com/zaproxy/zap-core-help

### Update your Remotes
Next, update the remotes for the repositories you cloned on your system.
1. Navigate to the cloned repository.
1. Point the remote *origin* to your repository.
1. Point the remote *upstream* to the main repository.

Here are the commands for two repositories; the procedure for others is identical. Replace `USERNAME` with your GitHub username.

```bash
cd zaproxy
git remote set-url origin https://github.com/USERNAME/zaproxy.git
git remote add upstream https://github.com/zaproxy/zaproxy.git
cd ..

cd zap-extensions
git remote set-url origin https://github.com/USERNAME/zap-extensions.git
git remote add upstream https://github.com/zaproxy/zap-extensions.git
cd ..
```

> ### Make changes and submit a PR!

Reviewer: ```suggestion
### Make Changes and Submit a PR!
```
